ref(rules): Refactor `get_rate` to be used for bulk queries

ceorourke opened the pull request:
The `apply_delayed` rule processor needs to call `get_rate` with bulk Snuba queries (see [this code block](https://github.com/getsentry/sentry/blob/088b9c73b604743864f76bd39a81a3238699d306/src/sentry/rules/processing/delayed_processing.py#L136-L171)), so this PR refactors `get_rate` to handle that.

ceorourke commented on src/sentry/rules/conditions/event_frequency.py:
```diff
@@ -208,28 +210,76 @@ def batch_query_hook(
+            result = percent_increase(result, comparison_result)
+        return result
 
+    def get_rate_bulk(
```
This is untested but also unused, the entire point of refactoring this is so I can use it in https://github.com/getsentry/sentry/pull/69167 which does / will have tests for this path.

saponifi3d replied:
Should we change the merge target to be `ceorourke/apply_delayed_rules` instead of `master` then? it might limit any risk of changing `get_rate` which is actively being used.

nhsiehgit replied:
If this is untested and unused, can we simply move it into #69167 instead?

saponifi3d commented on src/sentry/rules/conditions/event_frequency.py:
```diff
@@ -208,28 +210,78 @@ def batch_query_hook(
+        start = end - duration
+        return (start, end)
+
+    def get_comparison_start_end(
```
could we replace `get_start_end_from_duration` with `self.get_comparison_start_end(0, duration)`?

saponifi3d reviewed (approved):
generally lgtm, curious about the target branch change and your thoughts about `get_start_end_from_duration` though.

saponifi3d commented on src/sentry/rules/conditions/event_frequency.py:
```diff
@@ -139,8 +139,10 @@ def passes(self, event: GroupEvent, state: EventState) -> bool:
-        # TODO(mgaeta): Bug: Rule is optional.
-        current_value = self.get_rate(event, interval, self.rule.environment_id)  # type: ignore[arg-type, union-attr]
+        comparison_type = self.get_option("comparisonType", ComparisonType.COUNT)
+        comparison_interval = COMPARISON_INTERVALS[self.get_option("comparisonInterval")][1]
```
since this is a pretty deep lookup, should we wrap it / add a default value?

ceorourke commented:
@saponifi3d idk why I can't reply to your comment about the branch, I'd prefer to find out now rather than later when we have a lot more going on if I somehow fucked this up

nhsiehgit commented on src/sentry/rules/conditions/event_frequency.py:
```diff
@@ -208,28 +211,73 @@ def batch_query_hook(
+    def get_rate(
+        self,
+        duration: timedelta,
+        interval: timedelta,
```
can we name this `comparison_interval` to keep it consistent?

nhsiehgit commented on src/sentry/rules/conditions/event_frequency.py:
```diff
@@ -139,8 +139,11 @@ def passes(self, event: GroupEvent, state: EventState) -> bool:
+        comparison_interval_option = self.get_option("comparisonInterval", "5m")
+        comparison_interval = COMPARISON_INTERVALS[comparison_interval_option][1]
+        _, duration = self.intervals[interval]
+        current_value = self.get_rate(duration, comparison_interval, event, self.rule.environment_id, comparison_type)  # type: ignore[arg-type, union-attr]
```
nit - since we're passing a lot of values, can we pass these as kwargs instead?

nhsiehgit reviewed (commented):
So we've now added comparison type and interval to the data fixtures to determine our rates?

a few nit comments

ceorourke commented:
> So we've now added comparison type and interval to the data fixtures to determine our rates?
> 
> a few nit comments

no, nothing is added. I just moved stuff around
